src/base/MPSCQueue.hpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * MPSCQueue.hpp
+ */
+
+#ifndef _MPSCQueue_HPP_
+#define _MPSCQueue_HPP_
+
+#ident "$Id$"
+
+#include <atomic>
+#include <cassert>
+#include <optional>
+
+namespace cubbase
+{
+  template <typename T>
+  class MPSCQueue
+  {
+    private:
+      struct Node
+      {
+	Node (const T &data);
+
+	T m_data;
+	std::atomic<Node *> m_next;
+      };
+
+    public:
+      MPSCQueue ();
+      ~MPSCQueue ();
+
+      void enqueue (const T &item);
+      std::optional<T> dequeue ();
+      bool empty () const;
+
+    private:
+      std::atomic<Node *> m_head;
+      Node *m_tail;
```

Comment:
```suggestion
      alignas(64) std::atomic<Node *> m_head;
      alignas(64) Node *m_tail;
```

NIT: align 맞추면 좀 더 빨라지려나요?

Reply:
앗. MPSCQueue는 현재 사용되고 있지 않아서 제외하겠습니다. 이외에도 불필요한 부분들은 코드 정리가 조금 필요하겠네요.